File: csharp/RockPaperScissors.Tests/RockPaperScissorsTest.cs
namespace RockPaperScissors.Tests;

public class RockPaperScissorsTest
{
    [Theory]
    [InlineData("rock", "rock", 0)]
    [InlineData("paper", "paper", 0)]
    [InlineData("scissors", "scissors", 0)]
    [InlineData("rock", "scissors", -1)]
    [InlineData("scissors", "rock", 1)]
    [InlineData("paper", "rock", -1)]
    [InlineData("rock", "paper", 1)]
    [InlineData("scissors", "paper", -1)]
    [InlineData("paper", "scissors", 1)]
    public void TestWinner(string playerOne, string playerTwo, int expectedWinner)
    {
        // Arrange
        var game = new RockPaperScissors(playerOne, playerTwo);

        // Act
        var actualWinner = game.Winner();

        // Assert
        Assert.Equal(expectedWinner, actualWinner);
    }

    [Theory]
    [InlineData("", "")]
    [InlineData("ROCK", "paper")]
    [InlineData(" rock ", "paper")]
    [InlineData("sceezors", "rock")]
    [InlineData("scissors", "rocc")]
    [InlineData("paypa", "scissors")]
    [InlineData("paper", "5c1550r5")]
    public void TestValidation(string playerOne, string playerTwo)
    {
        Assert.Throws<ArgumentException>(() => new RockPaperScissors(playerOne, playerTwo));
    }
}

File: python/rock_paper_scissors/rock_paper_scissors_test.py
import pytest

from .rock_paper_scissors import RockPaperScissors


@pytest.mark.parametrize(
    "player_one,player_two,expected_winner",
    [
        ("rock", "rock", 0),
        ("paper", "paper", 0),
        ("scissors", "scissors", 0),
        ("rock", "scissors", -1),
        ("scissors", "rock", 1),
        ("paper", "rock", -1),
        ("rock", "paper", 1),
        ("scissors", "paper", -1),
        ("paper", "scissors", 1),
    ],
)
def test_winner(player_one, player_two, expected_winner):
    # Arrange
    game = RockPaperScissors(player_one, player_two)

    # Act
    actual_winner = game.winner()

    # Assert
    assert actual_winner == expected_winner


@pytest.mark.parametrize(
    "player_one,player_two",
    [
        ("", ""),
        ("ROCK", "paper"),
        (" rock ", "paper"),
        ("sceezors", "rock"),
        ("scissors", "rocc"),
        ("paypa", "scissors"),
        ("paper", "5c1550r5"),
    ],
)
def test_validation(player_one, player_two):
    with pytest.raises(ValueError) as e:
        RockPaperScissors(player_one, player_two)
    assert "invalid" in str(e)
